packages/playwright-test/src/reporters/azure.ts

```diff
@@ -0,0 +1,475 @@
+        this.attachmentsType = ['screenshot'];
+      } else {
+        this.attachmentsType = options.attachmentsType;
+      }
+    }
+
+    this.orgUrl = options.orgUrl;
+    this.projectName = options.projectName;
+    this.planId = options.planId;
+    this.token = options.token;
+    this.environment = options?.environment || undefined;
+    this.logging = !!options?.logging;
+    this.testRunTitle = `${this.environment ? `[${this.environment}]:` : ''} ${options?.testRunTitle || 'Playwright Test Run'}` ||
+      `${this.environment ? `[${this.environment}]:` : ''}Test plan ${this.planId}`;
+    this.uploadAttachments = options?.uploadAttachments || false;
+  }
+
+  async onBegin(): Promise<void> {
+    if (this.isDisabled)
+      return;
+
+    try {
+      this.connection = new azdev.WebApi(this.orgUrl, azdev.getPersonalAccessTokenHandler(this.token));
+      this.testApi = await this.connection.getTestApi();
+
+      const run = await this.createRun(this.testRunTitle);
+      if (run) {
+        this.runId = run.id;
+        this.log(colors.green(`Using run ${this.runId} to publish test results`));
+      } else {
+        this.isDisabled = true;
+      }
+    } catch (error) {
+      this.isDisabled = true;
+      this.warning(`Failed to create test run: ${error.message}. Check your token and orgUrl. Reporting is disabled.`);
+    }
+  }
+
+  async onTestEnd(test: TestCase, testResult: TestResult): Promise<void> {
+    await this.awaitForRunId();
+    if (this.isDisabled)
+      return;
+
+    this.logTestItem(test, testResult);
+    await this.publishCaseResult(test, testResult);
+  }
+
+  async onEnd(): Promise<void> {
+    await this.awaitForRunId();
+    if (this.isDisabled)
+      return;
+
+    let prevCount = this.resultsToBePublished.length;
+    while (this.resultsToBePublished.length > 0) {
+      // need wait all results to be published
+      if (prevCount > this.resultsToBePublished.length) {
+        this.log(
+            colors.gray(`Waiting for all results to be published. Remaining ${this.resultsToBePublished.length} results`)
+        );
+        prevCount--;
+      }
+      await new Promise(resolve => setTimeout(resolve, 250));
+    }
+
+    if (this.publishedResultsCount === 0 && !this.runId) {
+      this.log(colors.gray('No testcases were matched. Ensure that your tests are declared correctly.'));
+      return;
+    }
+
+    try {
+      if (!this.testApi)
+        this.testApi = await this.connection.getTestApi();
+      const runUpdatedResponse = await this.testApi.updateTestRun({ state: 'Completed' }, this.projectName, this.runId);
+      this.log(colors.green(`Run ${this.runId} - ${runUpdatedResponse.state}`));
+    } catch (err) {
+      this.log(`Error on completing run ${err as string}`);
+    }
+  }
+
+  printsToStdio() {
+    return true;
+  }
+
+  private log(message?: any) {
+    if (this.logging)
```

Comment:
I'm thinking if we should drop all that logging and use our common way of logging with the [`debug`](https://www.npmjs.com/package/debug) library, e.g. `pw:test:azure` as a key and document it.